### ISSUE TITLE
Snake case being ignored in an extended class (MethodNameInvalid)

### ISSUE THREAD

**Issue by psaikali** [WordPress-Coding-Standards/WordPress-Coding-Standards#1494]
The `MethodNameInvalid` does not work properly when extending a class.

### Scenario 1 (problematic)
```
class Snake_Case extends Something {
	public function snakeCase( $param ) {

	}
}
```
... does not trigger any error.

### Scenario 2 (working fine)
```
class Snake_Case {
	public function snakeCase( $param ) {

	}
}
```
... results in `Method name "snakeCase" in class Snake_Case is not in snake case format, try "snake_case" (WordPress.NamingConventions.ValidFunctionName.MethodNameInvalid)`

![Gif of the issue](https://dha4w82d62smt.cloudfront.net/items/111Y2e1d0v0X1q1n090F/Screen%20Recording%202018-09-29%20at%2002.16%20PM.gif?X-CloudApp-Visitor-Id=679133)

Using PHP 7.1.16 & PHPCS 3.3.1.

**Comment by GaryJones** [comment]
Thanks for reporting @psaikali.

This behaviour is (currently) by design:

https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/blob/1cb6a74615be51390776c3c22438c21b010c1737/WordPress/Sniffs/NamingConventions/ValidFunctionNameSniff.php#L139-L142

I guess it may be possible to check if a method of the same name exists in the parent class or interface, before ignoring it.

**Comment by jrfnl** [comment]
> I guess it may be possible to check if a method of the same name exists in the parent class or interface, before ignoring it.

Unfortunately not. For one, files are scanned in an order we can't influence, so the class being extended may be scanned _before_ or _after_ the extending class, which would make checking for this unreliable.

But more than that, the class being extended may not be scanned at all, like if a PHP native class is being extended or if a class is pulled in via Composer and put in the `vendor` directory - which would normally be excluded from PHPCS scans as dependencies don't have to comply with the same code style as your project-.

**Comment by GaryJones** [comment]
It would be worth adding that (excellent) explanation to the relevant point in the code :-)

**Comment by psaikali** [comment]
Thanks for the explanation @GaryJones ! 
Just out of curiosity, any reason why a child class could be allowed to use camelCase ? Not sure I get the reasoning here.

**Comment by aristath** [comment]
>  Just out of curiosity, any reason why a child class could be allowed to use camelCase ? Not sure I get the reasoning here.

As developers we sometimes extend 3rd-party PHP classes/libraries that are not designed with WordPress in mind and therefore don't nevessarily follow WordPress Coding Standards. If we want to override a method from the parent class then we have to follow the naming conventions of said parent class. So it makes sense not to throw errors in that case.

**Comment by psaikali** [comment]
Indeed, makes perfect sense, thanks for the explanation!
